pulp_ansible/app/downloaders.py

```diff
@@ -0,0 +1,74 @@
+from logging import getLogger
+import asyncio
+import backoff
+import json
+
+from aiohttp.client_exceptions import ClientResponseError
+
+from pulpcore.plugin.download import http_giveup, HttpDownloader
+
+
+log = getLogger(__name__)
+
+
+class TokenAuthHttpDownloader(HttpDownloader):
+    """
+    Custom Downloader that automatically handles Token Based and Basic Authentication.
+
+    Additionally, use custom headers from DeclarativeArtifact.extra_data['headers']
+    """
+
+    token_lock = asyncio.Lock()
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the downloader.
+        """
+        self.remote = kwargs.pop("remote")
+        super().__init__(*args, **kwargs)
+
+    @backoff.on_exception(backoff.expo, ClientResponseError, max_tries=10, giveup=http_giveup)
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        This method is decorated with a backoff-and-retry behavior to retry HTTP 429 errors. It
+        retries with exponential backoff 10 times before allowing a final exception to be raised.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+
+        """
+        if not self.remote.token:
+            return await super()._run(extra_data=extra_data)
+
+        token = await self.update_token()
+        headers = {"Authorization": "Bearer {token}".format(token=token)}
```

Comment:
I think this will fail if the CollectionRemote is pointing at galaxy.ansible.com  (and I believe, if pointed against a standalone galaxy_ng that is using drf style 'Authorization: Token <token here>' style auth headers)